Early engine failure throws NoSuchElementException and masks root exception

The only failure mode is:
```
org.junit.platform.commons.JUnitException: TestEngine with ID 'testng' failed to execute tests
Caused by: java.util.NoSuchElementException
```

Setting a debug breakpoint in `NoSuchElementException` shows the following.
```
"Test worker@1" tid=0x1 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
   at java.util.NoSuchElementException.<init>(NoSuchElementException.java:45)
   at java.util.Spliterators$1Adapter.next(Spliterators.java:688)
   at org.junit.support.testng.engine.ExecutionListener.chain(ExecutionListener.java:240)
   at org.junit.support.testng.engine.ExecutionListener.abortedOrFailed(ExecutionListener.java:230)
   at org.junit.support.testng.engine.ExecutionListener.toTestExecutionResult(ExecutionListener.java:225)
   at org.junit.support.testng.engine.ExecutionListener.lambda$onAfterClass$3(ExecutionListener.java:93)
   at org.junit.support.testng.engine.ExecutionListener$$Lambda/0x00007f88ac7b4248.accept(Unknown Source:-1)
   at org.junit.support.testng.engine.TestClassRegistry.lambda$finish$2(TestClassRegistry.java:53)
   at org.junit.support.testng.engine.TestClassRegistry$$Lambda/0x00007f88ac61b058.apply(Unknown Source:-1)
   at java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1940)
   - locked <0x241d> (a java.util.concurrent.ConcurrentHashMap$Node)
   at org.junit.support.testng.engine.TestClassRegistry.finish(TestClassRegistry.java:47)
   at org.junit.support.testng.engine.ExecutionListener.onAfterClass(ExecutionListener.java:89)
   at org.testng.internal.invokers.TestMethodWorker.invokeListenersOnAfterClass(TestMethodWorker.java:235)
   at org.testng.internal.invokers.TestMethodWorker.invokeAfterClassMethods(TestMethodWorker.java:214)
   at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:131)
... 
```

This indicates that `ExecutionListener.chain()` is calling `iterator.next()` blindly without guarding with an `if (iterator.hasNext()){...}`. The assumption of the caller (`abortedOrFailed()`) is that all aborted or failed results will have a throwable. It appears this logic is invoked not only for `@Test` methods but also for `@BeforeMethod` methods, and in this case a throwable is not available when there is an early engine failure (e.g. in `@BeforeSuite`, `@BeforeTest`, etc).

--- a/src/main/java/org/junit/support/testng/engine/ExecutionListener.java
+++ b/src/main/java/org/junit/support/testng/engine/ExecutionListener.java
@@ -237,9 +237,12 @@ class ExecutionListener extends DefaultListener {
 
 	private static Throwable chain(Stream<Throwable> failures) {
 		Iterator<Throwable> iterator = failures.iterator();
-		Throwable throwable = iterator.next();
-		iterator.forEachRemaining(throwable::addSuppressed);
-		return throwable;
+        Throwable throwable = null;
+        if (iterator.hasNext()) {
+            throwable = iterator.next();
+            iterator.forEachRemaining(throwable::addSuppressed);
+        }
+        return throwable;
 	}
 
 	static class MethodProgress {

--- a/src/main/java/org/junit/support/testng/engine/ExecutionListener.java
+++ b/src/main/java/org/junit/support/testng/engine/ExecutionListener.java
@@ -237,12 +237,12 @@ class ExecutionListener extends DefaultListener {
 
 	private static Throwable chain(Stream<Throwable> failures) {
 		Iterator<Throwable> iterator = failures.iterator();
-        Throwable throwable = null;
-        if (iterator.hasNext()) {
-            throwable = iterator.next();
-            iterator.forEachRemaining(throwable::addSuppressed);
-        }
-        return throwable;
+		Throwable throwable = null;
+		if (iterator.hasNext()) {
+			throwable = iterator.next();
+			iterator.forEachRemaining(throwable::addSuppressed);
+		}
+		return throwable;
 	}
 
 	static class MethodProgress {

--- a/src/test/java/org/junit/support/testng/engine/AbstractIntegrationTests.java
+++ b/src/test/java/org/junit/support/testng/engine/AbstractIntegrationTests.java
@@ -11,18 +11,24 @@
 package org.junit.support.testng.engine;
 
 import static java.util.function.Predicate.isEqual;
+import static org.assertj.core.api.Assertions.allOf;
 import static org.junit.platform.commons.util.FunctionUtils.where;
+import static org.junit.platform.engine.TestExecutionResult.Status.ABORTED;
 import static org.junit.platform.testkit.engine.Event.byTestDescriptor;
 import static org.junit.platform.testkit.engine.EventConditions.container;
 import static org.junit.platform.testkit.engine.EventConditions.displayName;
 import static org.junit.platform.testkit.engine.EventConditions.event;
+import static org.junit.platform.testkit.engine.EventConditions.finished;
 import static org.junit.platform.testkit.engine.EventConditions.uniqueIdSubstring;
+import static org.junit.platform.testkit.engine.TestExecutionResultConditions.status;
 
 import java.nio.file.Path;
+import java.util.Optional;
 
 import org.assertj.core.api.Condition;
 import org.junit.jupiter.api.io.TempDir;
 import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestExecutionResult;
 import org.junit.platform.testkit.engine.EngineTestKit;
 import org.junit.platform.testkit.engine.Event;
 
@@ -48,4 +54,13 @@ abstract class AbstractIntegrationTests {
 			byTestDescriptor(where(TestDescriptor::getLegacyReportingName, isEqual(legacyReportingName))),
 			"descriptor with legacy reporting name '%s'", legacyReportingName);
 	}
+
+	static Condition<Event> abortedWithoutReason() {
+		return finished(allOf(status(ABORTED), emptyThrowable()));
+	}
+
+	static Condition<TestExecutionResult> emptyThrowable() {
+		return new Condition<>(where(TestExecutionResult::getThrowable, Optional::isEmpty), "throwable is empty");
+	}
+
 }

--- a/src/test/java/org/junit/support/testng/engine/ConfigurationMethodIntegrationTests.java
+++ b/src/test/java/org/junit/support/testng/engine/ConfigurationMethodIntegrationTests.java
@@ -10,6 +10,7 @@
 
 package org.junit.support.testng.engine;
 
+import static org.assertj.core.api.Assertions.allOf;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.platform.engine.discovery.DiscoverySelectors.selectClass;
 import static org.junit.platform.testkit.engine.EventConditions.abortedWithReason;
@@ -74,7 +75,7 @@ class ConfigurationMethodIntegrationTests extends AbstractIntegrationTests {
 			event(testClass(testClass), started()), //
 			event(test("method:test()"), started()), //
 			event(test("method:test()"), abortedWithReason(message("boom"))), //
-			event(testClass(testClass), abortedWithReason()), //
+			event(testClass(testClass), abortedWithoutReason()), //
 			event(engine(), finishedWithFailure(message("boom"))));
 	}
 

--- a/src/test/java/org/junit/support/testng/engine/ConfigurationMethodIntegrationTests.java
+++ b/src/test/java/org/junit/support/testng/engine/ConfigurationMethodIntegrationTests.java
@@ -10,7 +10,6 @@
 
 package org.junit.support.testng.engine;
 
-import static org.assertj.core.api.Assertions.allOf;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.platform.engine.discovery.DiscoverySelectors.selectClass;
 import static org.junit.platform.testkit.engine.EventConditions.abortedWithReason;

--- a/src/test/java/org/junit/support/testng/engine/ConfigurationMethodIntegrationTests.java
+++ b/src/test/java/org/junit/support/testng/engine/ConfigurationMethodIntegrationTests.java
@@ -74,7 +74,7 @@ class ConfigurationMethodIntegrationTests extends AbstractIntegrationTests {
 			event(testClass(testClass), started()), //
 			event(test("method:test()"), started()), //
 			event(test("method:test()"), abortedWithReason(message("boom"))), //
-			event(testClass(testClass), finishedSuccessfully()), //
+			event(testClass(testClass), abortedWithReason()), //
 			event(engine(), finishedWithFailure(message("boom"))));
 	}
 

--- a/src/testFixtures/java/example/configuration/methods/FailingBeforeSuiteConfigurationMethodTestCase.java
+++ b/src/testFixtures/java/example/configuration/methods/FailingBeforeSuiteConfigurationMethodTestCase.java
@@ -10,6 +10,7 @@
 
 package example.configuration.methods;
 
+import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.BeforeSuite;
 import org.testng.annotations.Test;
 
@@ -18,6 +19,11 @@ public class FailingBeforeSuiteConfigurationMethodTestCase {
 	@BeforeSuite
 	public void beforeSuite() {
 		throw new AssertionError("boom");
+	}
+
+	@BeforeMethod
+	public void beforeMethod() {
+		// never called
 	}
 
 	@Test

--- a/src/testFixtures/java/example/configuration/methods/FailingBeforeTestConfigurationMethodTestCase.java
+++ b/src/testFixtures/java/example/configuration/methods/FailingBeforeTestConfigurationMethodTestCase.java
@@ -10,6 +10,7 @@
 
 package example.configuration.methods;
 
+import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.BeforeTest;
 import org.testng.annotations.Test;
 
@@ -18,6 +19,11 @@ public class FailingBeforeTestConfigurationMethodTestCase {
 	@BeforeTest
 	public void beforeTest() {
 		throw new AssertionError("boom");
+	}
+
+	@BeforeMethod
+	public void beforeMethod() {
+		// never called
 	}
 
 	@Test

